File: algorithm-design/java/src/test/java/leetcode/LC001TwoSum.java
package leetcode;

import java.util.HashMap;
import java.util.Map;

import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

/**
 * https://leetcode.com/problems/two-sum/
 */
public class LC001TwoSum {
    public class Solution {
        public int[] twoSum(int[] nums, int target) {
            if (nums == null) {
                return null;
            }
            Map<Integer, Integer> map = new HashMap<>();
            for (int i = 0; i < nums.length; i++) {
                int expected = target - nums[i];
                Integer found = map.get(expected);
                if (found != null) {
                    return new int[]{found + 1, i + 1};
                } else {
                    map.put(nums[i], i);
                }
            }
            return null;
        }
    }

    @Test
    public void test_example() throws Exception {
        int[] nums = {2, 7, 11, 15};
        int[] expected = {1, 2};
        assertArrayEquals(expected, new Solution().twoSum(nums, 9));
    }
}

File: algorithms/leetcode/lc001_two_sum/LC001TwoSum.cs
// https://leetcode.com/problems/two-sum/

using System.Collections.Generic;

namespace AlgorithmDesign.leetcode.lc001_two_sum
{
    public class Solution
    {
        public int[] TwoSum(int[] nums, int target)
        {
            if (nums == null)
            {
                return null;
            }
            Dictionary<int, int> map = new Dictionary<int, int>();
            for (int i = 0; i < nums.Length; i++)
            {
                int expected = target - nums[i];
                if (map.ContainsKey(expected))
                {
                    int found = map[expected];
                    return new[] { found, i };
                }
                else
                {
                    map[nums[i]] = i;
                }
            }
            return null;
        }
    }
}
